File: src/api/new-discussion.ts
import { RequestFN } from "~/core/request-function";
import { EntityState, NewDiscussionRecipient, SessionHandle, TabLocation } from "~/models";

/**
 * Create a discussion.
 *
 * Sadly, we can't get the ID of the created discussion
 * or anything else related to it, you need to request the
 * discussions list once again.
*/
export const newDiscussion = async (
  session: SessionHandle,
  subject: string,
  content: string,
  recipients: Array<NewDiscussionRecipient>
): Promise<void> => {
  const request = new RequestFN(session, "SaisieMessage", {
    _Signature_: { onglet: TabLocation.Discussions },

    donnees: {
      contenu: session.user.authorizations.hasAdvancedDiscussionEditor ? {
        _T: 21,
        V: content
      } : content,

      objet: subject,
      estCreationCarnetLiaison: false,
      listeFichiers: [],
      listeDestinataires: recipients.map((recipient) => ({
        E: EntityState.MODIFICATION,
        G: recipient.kind,
        N: recipient.id
      }))
    }
  });

  await request.send();
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/new-discussion.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.models.EntityState
import ink.literate.pawnote.models.NewDiscussionRecipient
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/**
 * Create a discussion.
 *
 * Sadly, we can't get the ID of the created discussion or anything else related to it, you need to
 * request the discussions list once again.
 */
suspend fun newDiscussion(
    session: SessionHandle,
    subject: String,
    content: String,
    recipients: List<NewDiscussionRecipient>
) {
  val request =
      RequestFN(
          session.information,
          "SaisieMessage",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Discussions.code) }

                putJsonObject("donnees") {
                  if (session.user.authorizations.hasAdvancedDiscussionEditor)
                      putJsonObject("contenu") {
                        put("_T", 21)
                        put("V", content)
                      }
                  else put("contenu", content)

                  put("objet", subject)
                  put("estCreationCarnetLiaison", false)
                  putJsonArray("listeFichiers") {}
                  putJsonArray("listeDestinataires") {
                    for (recipient in recipients) addJsonObject {
                      put("E", EntityState.MODIFICATION.code)
                      put("G", recipient.kind.code)
                      put("N", recipient.id)
                    }
                  }
                }
              }))

  request.send()
}
